### PR TITLE
try to stop sam from hammering google during quota issues

### PR DESCRIPTION
**Pull request by dvoet**
Ticket: https://broadworkbench.atlassian.net/browse/ID-446

There are multiple instances of sam each with multiple threads doing google group things. When there is a quota error they all start hammering google with uncoordinated backoffs.

This tries to stop that using a database record saving the last time a quota error was encountered. If we want to make a google groups request within a configurable time (2s by default) after the last error, just don't do it and pretend kind of like we did and got the rate limit error. A little quick and dirty, it does so by overriding the `retryExponentially` function in the `HttpGoogleDirectoryDAO` to wrap operations with this desired feature.

---

**PR checklist**

- [ ] I've followed [the instructions](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've filled out the [Security Risk Assessment](https://sdarq.dsp-appsec.broadinstitute.org/jira-ticket-risk-assesment) (requires Broad Internal network access) and attached the result to the JIRA ticket

### CODE REVIEW

File: src/test/resources/reference.conf [hunk 1]
```diff
@@ -158,7 +158,7 @@ db {
 }
 
 scalikejdbc.global.loggingSQLAndTime {
-  enabled = false, # switch this to true to print sql
+  enabled = true, # switch this to true to print sql
```
Comment (tlangs) [inline src/test/resources/reference.conf]: Do we still want this?

Comment (dvoet) [reply src/test/resources/reference.conf]: nope, reverted

Comment (dvoet) [reply src/test/resources/reference.conf]: nope, reverted

**Review by tlangs — APPROVED** [review]

File: src/main/resources/org/broadinstitute/dsde/sam/liquibase/changesets/20230203_last_quota_error.xml [hunk 2]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<databaseChangeLog logicalFilePath="dummy"
+                   xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+                   xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext"
+                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.4.xsd">
+
+    <changeSet logicalFilePath="dummy" author="dvoet" id="last_quota_error">
+        <createTable tableName="SAM_LAST_QUOTA_ERROR">
```
Comment (gpolumbo-broad) [inline src/main/resources/org/broadinstitute/dsde/sam/liquibase/changesets/20230203_last_quota_error.xml:9]: nit - "last quota error" is a little generic, there are lots of quotas.  Ideally this table would be named more precisely or better yet, maybe there's an additional column for `QUOTA_TYPE` or `QUOTA_NAME`.  We can slay that dragon when it comes for our treasure though.

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/google/CoordinatedBackoffHttpGoogleDirectoryDAOSpec.scala [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        val googleDao = new CoordinatedBackoffHttpGoogleDirectoryDAO("", null, "", lastQuotaErrorDAOMock)
+
+        val test = googleDao.retryExponentially(GoogleUtilities.RetryPredicates.whenUsageLimited) { () =>
+          Future(fail("op should not have been called"))
+        }
+
+        Await.result(test, Duration.Inf) match {
+          case Right(_) => fail("expected error")
+          case Left(errors) =>
+            errors.length shouldBe 7
```
Comment (gpolumbo-broad) [inline src/test/scala/org/broadinstitute/dsde/workbench/sam/google/CoordinatedBackoffHttpGoogleDirectoryDAOSpec.scala:34]: What is this 7?

**Review by gpolumbo-broad — APPROVED** [review]

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/google/CoordinatedBackoffHttpGoogleDirectoryDAOSpec.scala (hunk 3 above)
Comment (dvoet) [reply src/test/scala/org/broadinstitute/dsde/workbench/sam/google/CoordinatedBackoffHttpGoogleDirectoryDAOSpec.scala:34]: my lucky number?

Comment (dvoet) [reply src/test/scala/org/broadinstitute/dsde/workbench/sam/google/CoordinatedBackoffHttpGoogleDirectoryDAOSpec.scala:34]: it is the number of times `retryExponentially` attempts the operation

**Pull request merged**